### PR TITLE
feat: mock resources

### PR DESCRIPTION
**Pull request by jackkav**
### goal: 
Support API mocking features.

### terminology: 
 - "mock server" a group of "mock routes", perhaps with a remotely assigned url the user can copy and use in their app, TBD as server provisioning is an external concern, perhaps mock resource
 - "mock route"  a stubbed response, perhaps with remote url, route to resolve, TBD as route clashes with react route
 - "example response" a saved template used to populate a request bin, saved from a custom form or generated from an OAS spec or existing response

### ideas for next pass:
- could transform a request collection into a mock server/mock routes
- could transform a design document into a mock server/mock routes
- could add a mock server navigation to design document menu pills and add pills to collection
- could use kong dialect to describe services and routes if we can find a payoff in orchestration or extension
- should derived mock servers be editable?
- could use the same relationships as a spec, however scope will be migrated if old version is used

### questions
- data model
  - [x] to use mock route path and name or just name?
    - additive, it should be cheap to add. apth field later if we need to use name for something else
  - [x] to record url in the route or make it a product of server
    - remove url from route
  - [x] to write response body to file in order to model responses more closely
    - could create body size and content type validator to ensure responses are always human readable content
 - should be a child of project or workspace?
- request bins cannot be edited, should we reuse or rebuild?
- should mock servers interact with environments?
- abstractions
  - requestheaderseditor
- alternate protocols? besides http, sse, gql, grpc, websockets?

### todo:
- [x] project pane UI
- [x] wire up mock server crud actions
- [x] wire up mock route crud actions
- [x] sidebar
- [x] url bar
- [x] response editor
- [x] mock response UI inside collection
- [x] send to mock url feature, insomnia network stack side, show response, from request tab and from mock server editor
- [x] request bin/mock server activity fetched from mockbin api
- [x] send to mock url, server POC side
- [x] path and naming?
- [x] fix command palette
- [x] remove offline mode workaround
- [x] fix button sizing of collection settings
- [x] icon bg color
- [x] in order not to break downgraded mock servers we will need to make a new workspace key to replace scope, because scope is forced to defualt to design when unknown prior to adding mock server.
- [x] fix duplicate
- [x] rename mock server to mock
- [x] fix mock settings
- [x] persist mock route onblur in tabs
- [x] default cloud mock server, radio button local mock server and input, use create project dialog style
- [x] check route modal for unused fields
- [x] sync
- [x] onblur update mockbin on header key or value blur if key has value
- [x] smoke test
  - github action containerised services for redis and mockbin
  - use our infra
  - build fs adapter and run mockbin in the action with node
- [x] move mock response tab into response pane
- [x] maybe move url out of mock route and make it just a product of server id and env var
- [x] maybe write response body to file
- [x] create only one hidden request under mock route

### future work

- [ ] spike transform OAS response objects to mock server
- [ ] aggregate logs per server somehow
- [ ] sync with mockbin at route patch time rather than seperately, handle errors differently
- [ ] sync story and nesting mocks inside desing/debug/test/mock thing
- [ ] extract the path from a url when extracting a mock from a response
- [ ] extract multiple mocks from collection/folder by paths and responses
- [ ] extract from HARs
- [ ] folders and drag and drop, ordering
- [ ] restrict mock body size in order to avoid breaking the database
- [ ] improve extract mock ux, 
  - [ ] sync with mockbin after extraction
  - [ ] infer path from request url
  - [ ] block oversized and invalid content type
- [ ] use insomniaFetch with url override?
- [ ] auto add content type header?

### design:

image

- should history be the default pane?
- mockbin api cannot specify a method, still show it for test button?
- does environment make sense to use for this?
- does cookie modal make sense here?

### missteps

- made a new mock server object and try to show it on the project view, too difficult to play nice with the workspaceWithMetadata concept, easier to match design spec -> workspace relationship

### Notes

attempting to wire up without pathing server to get a working pass before adding a dependency on a backend store for resolving routes to bins.
this begs a few questions; what if the backend is inconsistent with the ui what the user experience should be? what if the backend is inconsistent with the mockbin state.

Ideally all we will have access to the mockbin state store to avoid inconsistencies.

closes INS-3246

### CODE REVIEW

**Comment by jackkav** [comment]
image

**Review by filfreire — DISMISSED** [review]
🏎️🏎️

File: packages/insomnia/src/common/constants.ts [hunk 1]
```diff
@@ -136,6 +136,7 @@ export enum UpdateURL {
 
 // API
 export const getApiBaseURL = () => env.INSOMNIA_API_URL || 'https://api.insomnia.rest';
+export const getMockServiceURL = () => env.INSOMNIA_MOCK_API_URL || 'http://localhost:8080';
```
Comment (jackkav) [inline packages/insomnia/src/common/constants.ts]: point this at cloud

File: packages/insomnia/src/ui/components/editors/mock-response-extractor.tsx [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        onSubmit={async e => {
+          e.preventDefault();
+          if (!selectedMockServer || !selectedMockRoute) {
+            return;
+          }
+
+          if (activeResponse) {
+            // TODO: move this out of the renderer, and upsert mock
+            const body = await fs.readFile(activeResponse.bodyPath);
+            console.log('Exporting response to mock route', activeResponse, body.toString());
```
Comment (jackkav) [inline packages/insomnia/src/ui/components/editors/mock-response-extractor.tsx]: remove comment

File: packages/insomnia/src/ui/components/editors/mock-response-extractor.tsx [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+            patchMockRoute(selectedMockRoute, {
+              body: body.toString(),
+              mimeType: activeResponse.contentType,
+              statusCode: activeResponse.statusCode,
+              headers: activeResponse.headers,
+            });
```
Comment (jackkav) [inline packages/insomnia/src/ui/components/editors/mock-response-extractor.tsx:46]: experiment moving the upsert into the mock route patch in order to keep synced.

File: packages/insomnia/src/ui/routes/mock-route.tsx [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+  };
+};
+
+const mockContentTypes = [
+  CONTENT_TYPE_PLAINTEXT,
+  CONTENT_TYPE_JSON,
+  CONTENT_TYPE_XML,
+  CONTENT_TYPE_YAML,
+];
+export const formToHar = ({ statusCode, statusText, mimeType, headersArray, body }: { statusCode: number; statusText: string; mimeType: string; headersArray: RequestHeader[]; body: string }): HarResponse => {
```
Comment (jackkav) [inline packages/insomnia/src/ui/routes/mock-route.tsx]: rename to mockRouteToHar

File: packages/insomnia/src/ui/routes/mock-route.tsx [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+  const createandSendPrivateRequest = (patch: Partial<Request>) =>
+    requestFetcher.submit(JSON.stringify(patch),
+      {
+        encType: 'application/json',
+        action: `/organization/${organizationId}/project/${projectId}/workspace/${workspaceId}/debug/request/new-mock-send`,
+        method: 'post',
+      });
+
+  const upsertMockbinHar = async (pathInput?: string) => {
+    console.log('upserting mockbin har');
```
Comment (jackkav) [inline packages/insomnia/src/ui/routes/mock-route.tsx]: remove console logs

File: packages/insomnia-smoke-test/tests/smoke/mock.test.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { test } from '../../playwright/test';
+
+// TODO: unskip this test when cloud mock is online
+test.skip('can make a mock route', async ({ page }) => {
```
Comment (gatzjames) [inline packages/insomnia-smoke-test/tests/smoke/mock.test.ts:4]: Remember to change this when the api is deployed

File: packages/insomnia/src/ui/components/modals/workspace-settings-modal.tsx [hunk 7]
```diff
@@ -63,45 +74,97 @@ export const WorkspaceSettingsModal = ({ workspace, onClose }: Props) => {
...
+                      placeholder="Write a description"
+                      defaultValue={workspace.description}
+                      onChange={(description: string) => {
+                        workspacePatcher(workspace._id, { description });
+                      }}
+                    />
+                    <Heading>Actions</Heading>
+                    <PromptButton
+                      onClick={async () => {
+                        const docs = await db.withDescendants(workspace, models.request.type);
```
Comment (gatzjames) [inline packages/insomnia/src/ui/components/modals/workspace-settings-modal.tsx:106]: @TODO: For next pass move this to database methods and perhaps put it on an action

**Review by gatzjames — DISMISSED** [review]

**Review by filfreire — APPROVED** [review]

**Pull request merged**